### PR TITLE
Migrate SMS Pillow

### PR DESCRIPTION
**Pull request by gcapalbo**
@czue 
This is still a work in progress but I figured I'd open for review.  There's still a bug I'm working out, but overall I've tested that it publishes changes to Kafka and that the new SqlSMSPillow consumes the changes and syncs the docs to Elasticsearch.

In terms of how to release this, I was thinking that in a first iteration we'd first run the publish part and not the SqlSMSPillow.  I'll then run a management command to publish a change to Kafka for each SMS in the DB and set the checkpoint for the new SqlSMSPillow to be the last change.  Then we can turn on the SqlSMSPillow and it will have just a small amount of catching up to do.

That way, the Kafka feed will always have the whole history in case we need to reprocess all changes. Does that sound right, and do you forsee any issues with running that management command which will publish a few million changes all at once (they shouldn't be consumed by anything since SMS has its own topic).

cc: @benrudolph

### CODE REVIEW

**Comment by czue** [comment]
@gcapalbo sorry will prioritize taking a look at this tomorrow

File: corehq/apps/sms/models.py [hunk 1]
```diff
@@ -498,16 +498,30 @@ def outbound_backend(self):
 
 
 class SMS(SMSBase):
+    def to_json(self):
```
Comment (benrudolph) [inline corehq/apps/sms/models.py]: fyi, we have django rest framework installed now which makes serialization very easy. here's an example: https://github.com/dimagi/commcare-hq/blob/master/corehq/form_processor/serializers.py

File: corehq/pillows/sms.py [hunk 2]
```diff
@@ -42,3 +55,22 @@ def calc_meta(cls):
         so we just do a hash of the "prototype" instead to determined md5
         """
         return cls.calc_mapping_hash({"es_meta": cls.es_meta, "mapping": cls.default_mapping})
+
+
+def get_sql_sms_pillow():
```
Comment (czue) [inline corehq/pillows/sms.py]: You'll need to add an ID to this now that https://github.com/dimagi/commcare-hq/pull/10816 has been merged.

**Comment by czue** [comment]
looking good! I think we should definitely add at least one integration test to make sure the end-to-end setup works. You could use this as an example: https://github.com/dimagi/commcare-hq/blob/master/testapps/test_pillowtop/tests/test_case_pillow.py#L77

File: corehq/apps/sms/models.py (hunk 1 above)
Comment (gcapalbo) [inline corehq/apps/sms/models.py]: cool, will take a look at that

**Comment by gcapalbo** [comment]
@czue thanks for the feedback - will update.  does my approach on how to release this sound good (see pr description)?

**Comment by czue** [comment]
Oh thanks, meant to respond to that.

So Kafka does have a persistent log that we could keep indefinitely, but currently changes older than 7 days are purged, and I would prefer to keep it that way for space/scale considerations. So reindexing needs to be done via an alternate mechanism (typically manually iterating through a DB query and sending the results to a pillow). So I think it would be better if we create a reindexer for this that doesn't depend on everything being in the kafka feed if and when we need it. I had started to sketch out how I think this could work in a way that's much lighter weight than what we current have [here](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/hqcase/management/commands/ptop_reindexer_v2.py#L6).

For releasing what I did with UCR was start publishing the changes and put the pillow logic behind a toggle (so the main pillow checks the toggle and skips if it's set, and the new pillow only processes if it's set). Then once I was happy with testing I deleted that logic and flipped the toggle entirely over to the new pillow. That could likely work here as well.

**Comment by gcapalbo** [comment]
@czue thanks, that makes sense - so sounds like for now i won't worry about making a reindexer and we'll do that when we need to.  i like the approach of using a toggle to put this into prod and test it - i'll do that here too.

another thing i wanted to ask - do new topics get automatically created on first publish? it seemed to automatically create the new topic for me locally, but only after the first change was published (the first publish change failed, then subsequent ones worked), but maybe i just had something else wrong initially and i didn't realize it.

**Comment by czue** [comment]
great. sounds right on the reindexer.

for creating the topic what you said is right. the first publish fails and creates it, and subsequent ones work. I have [a backlog ticket](http://manage.dimagi.com/default.asp?217905) to fix this in the consumers but I haven't gotten around to it yet. You can also just manually run a command to create them.

File: testapps/test_pillowtop/tests/test_sms_pillow.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
+from corehq.apps.change_feed import topics
+from corehq.apps.change_feed.consumer.feed import change_meta_from_kafka_message
+from corehq.apps.es.sms import SMSES
+from corehq.apps.sms.models import MessagingEvent, MessagingSubEvent, SMS
+from corehq.elastic import get_es_new
+from corehq.pillows.sms import get_sql_sms_pillow, ES_SMS_INDEX
+from corehq.util.elastic import delete_es_index, ensure_index_deleted
```
Comment (dimagimon) [inline testapps/test_pillowtop/tests/test_sms_pillow.py]: F401 'delete_es_index' imported but unused

File: testapps/test_pillowtop/tests/test_sms_pillow.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+from datetime import datetime
+from dimagi.utils.parsing import json_format_datetime
+from django.test import TestCase
+from mock import patch
```
Comment (dimagimon) [inline testapps/test_pillowtop/tests/test_sms_pillow.py]: F401 'patch' imported but unused

**Comment by gcapalbo** [comment]
@czue ok there's just two things left on this and then it's ready to go:

- do you know if I need to update https://github.com/dimagi/commcare-hq/blob/migrate-sms-pillow/corehq/apps/change_feed/topics.py#L19? that doesn't seem to be called from the new pillow

- the build seems to be failing because the new topic doesn't exist (but the test passes locally). before i dig deeper in how to create the topic in the test environment, is this a problem you've already solved anywhere?

**Comment by czue** [comment]
for the first question - that's only used in the pillow that streams couch models to kafka, so since these are coming from SQL it probably isn't necessary.

for the second question you just need to add the topic [here](https://github.com/dimagi/commcare-hq/blob/master/docker/utils.sh#L35)

**Comment by gcapalbo** [comment]
@czue great, thanks. once the build passes, i think we're good to go

File: corehq/pillows/sms.py [hunk 5]
```diff
@@ -42,3 +56,36 @@ def calc_meta(cls):
         so we just do a hash of the "prototype" instead to determined md5
         """
         return cls.calc_mapping_hash({"es_meta": cls.es_meta, "mapping": cls.default_mapping})
+
+    def change_transport(self, doc_dict):
```
Comment (gcapalbo) [inline corehq/pillows/sms.py]: @czue this is the right place in the couch SMSPillow to patch for the toggle right? i tested and it works but just want to doublecheck i didn't miss something

File: settings.py [hunk 6]
```diff
@@ -1440,6 +1440,11 @@
             'class': 'pillowtop.pillow.interface.ConstructedPillow',
             'instance': 'corehq.pillows.case.get_sql_case_to_elasticsearch_pillow',
         },
+        {
```
Comment (gcapalbo) [inline settings.py]: @czue is there anywhere else i need to add this for it to run in production? i figured i'd leave it under experimental for now and then move it to core once i remove the toggle

Comment (czue) [inline settings.py]: nope this should be it

File: corehq/pillows/sms.py (hunk 5 above)
Comment (czue) [inline corehq/pillows/sms.py]: yep, this looks good

File: testapps/test_pillowtop/tests/test_sms_pillow.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+            form_name=None,
+            xforms_session=None,
+            case_id=None,
+            status=MessagingEvent.STATUS_COMPLETED,
+            error_code=None,
+            additional_error_text=None
+        )
+        # Some of the values here don't apply for a simple outgoing SMS,
+        # but the point of this is to just test the serialization and that
+        # everything makes it to elasticsearch
```
Comment (czue) [inline testapps/test_pillowtop/tests/test_sms_pillow.py]: :+1:

**Comment by czue** [comment]
:boom:

**Pull request merged**